pkg/operator/ceph/cluster/watcher.go

```diff
@@ -130,6 +145,333 @@ func (c *clientCluster) onK8sNode(ctx context.Context, object runtime.Object) bo
+		err := c.fenceNode(ctx, node, cluster)
+		if err != nil {
+			return pkgerror.Wrapf(err, "failed to create network fence for node %q.", node.Name)
+		}
+		return nil
+	}
+
+	logger.Infof("node %s does not have taint %s, trying to remove networkFence CR if exists", node.Name, v1.TaintNodeOutOfService)
+	err = c.unfenceAndDeleteNetworkFence(ctx, *node, cluster)
+	if err != nil {
+		return pkgerror.Wrapf(err, "failed to delete network fence for node %q.", node.Name)
+	}
+
+	return nil
+}
+
+func (c *clientCluster) fenceNode(ctx context.Context, node *corev1.Node, cluster *cephv1.CephCluster) error {
+	volumesInuse := node.Status.VolumesInUse
+	if len(volumesInuse) == 0 {
+		logger.Debugf("no volumes in use for node %q", node.Name)
+		return nil
+	}
+	logger.Debugf("volumesInuse %s", volumesInuse)
+
+	rbdVolumesInUse, _ := getCephVolumesInUse(cluster, volumesInuse)
+	if len(rbdVolumesInUse) == 0 {
+		logger.Infof("no rbd `volumesInUse` for node %q", node.Name)
+		return nil
+	}
+
+	listPVs, err := c.context.Clientset.CoreV1().PersistentVolumes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return pkgerror.Wrapf(err, "failed to list PV")
+	}
+
+	rbdPVList, _ := listRBDAndCephFSPV(listPVs, cluster, rbdVolumesInUse)
+	if len(rbdPVList) == 0 {
+		logger.Info("No rbd PVs found on the node")
+		return nil
+	}
+
+	clusterInfo := cephclient.AdminClusterInfo(ctx, cluster.Namespace, cluster.Name)
+
+	// We only need to create the network fence for any one of rbd/cephFS pv.
+	err = c.fenceRbdImage(ctx, node, cluster, clusterInfo, rbdPVList[0])
+	if err != nil {
+		return pkgerror.Wrapf(err, "failed to fence rbd volumes")
+	}
+
+	// else {
+	// 	err = c.fennceCephFSVolumes(ctx, node, cluster, clusterInfo, listCephFSPV[0])
+	// 	if err != nil {
+	// 		return fmt.Errorf("failed to fence cephfs volumes. %v", err)
+	// 	}
+	// }
+
+	return nil
+}
+
+func getCephVolumesInUse(cluster *cephv1.CephCluster, volumesInUse []v1.UniqueVolumeName) ([]string, []string) {
+	var rbdVolumesInUse, cephFSVolumeInUse []string
+
+	for _, volume := range volumesInUse {
+		volumesInuseRemoveK8sPrefix := strings.TrimPrefix(string(volume), "kubernetes.io/csi/")
+		splitVolumeInUseBased := strings.Split(volumesInuseRemoveK8sPrefix, "^")
+		logger.Infof("volumeInUse after split based on '^' %v", splitVolumeInUseBased)
+
+		if len(splitVolumeInUseBased) == 2 && splitVolumeInUseBased[0] == fmt.Sprintf("%s.rbd.csi.ceph.com", cluster.Namespace) {
+			rbdVolumesInUse = append(rbdVolumesInUse, splitVolumeInUseBased[1])
+
+		}
+		// else if len(splitVolumeInUseBased) == 2 && splitVolumeInUseBased[0] == fmt.Sprintf("%s.cephfs.csi.ceph.com", cluster.Namespace) {
+		// 	cephFSVolumeInUse = append(cephFSVolumeInUse, splitVolumeInUseBased[1])
+		// }
+	}
+	return rbdVolumesInUse, cephFSVolumeInUse
+}
+
+func listRBDAndCephFSPV(listPVs *corev1.PersistentVolumeList, cluster *cephv1.CephCluster, rbdVolumesInUse []string) ([]corev1.PersistentVolume, []corev1.PersistentVolume) {
```

Comment:
Until we have the cephfs implementation, let's remove it from the method signature as well
```suggestion
func listRBDAndCephFSPV(listPVs *corev1.PersistentVolumeList, cluster *cephv1.CephCluster, rbdVolumesInUse []string) ([]corev1.PersistentVolume) {
```